aicsimageio/dimensions.py

```diff
@@ -142,3 +142,9 @@ def __getitem__(self, key: Union[str, Sequence[str]]) -> Tuple[int, ...]:
             raise TypeError(
                 f"Key must be a string or list of strings but got type {type(key)}"
             )
+
+    def __setattr__(self, __name: str, __value: int) -> None:
+        super().__setattr__(__name, __value)
+
+    def __getattribute__(self, __name: str) -> int:
```

Comment:
Entirely an FYI in case you haven't seen it yet: there's a great section in "Fluent Python" (2nd ed; Ch 12, section "Vector Take 3: Dynamic Attribute Access"; available online via SPL for free) about implementing a class with dynamic attribute access. The author doesn't include typings, but might be interesting to compare that implementation with this! Notable differences: the author makes use of `__getattr__` instead of `__getattribute__`, and defines and includes a check on an "allow list" of attributes within a `__match_args__` class attribute (a Python 3.10 thing which additionally opens up opportunities to make use of the class in pattern matching, but no harm in prior versions of Py).

Reply:
Huh. I tried `__getattr__` last night and got a weird error but tried it just now and it worked just fine. 🤷‍♀️ Will swag.

I likely won't handle `__match_args__`. Maybe in 5.0

Reply:
> It never uses this code, but ... Mypy can at least check it

Yep, that's my understanding of how `__getattr__` works here!  One minor thing now that you've updated to use `__getattr__` is that--assuming I'm looking at the right diff--you should probably update the [`super()` delegation](https://github.com/AllenCellModeling/aicsimageio/pull/419/commits/ee54793e03033c5db611af5479e6f88b69bd1657#diff-cd5efab4cdf319cce520f5f50ec06f2414b8d555f4321d59bcc2578eaeb32c82L150) to use `__getattr__` too?

Reply:
Totally correct!

Reply:
yeah `__getattr__` isn't define on everything.  if you want to override attribute access, a good pattern is to use `__getattr__`, but then call object.__getattribute__ as a fallback

```python
def __getattr__(self, name):
    ...  # go nuts
    return object.__getattribute__(self, name)
```

Reply:
I think the only question on my end is if I should make the return type all the possible types from Dimensions? `Union[str, int, Tuple[int]]` or just `int`.

Reply:
at first glance, since it looks like all of the internal usages of `setattr` are setting attributes to `int` ... `int` seems correct to me.  but when might you be returning the others?

Reply:
I was thinking of the `dims.order` and `dims.shape` but you are correct that the only things that `setattr` are `int` typed!